### PR TITLE
Framework: Created 3 new selectors (getNetworkSites, isConnectedSecondaryNetworkSite...

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Framework: Created 3 new selectors (isMainSiteOf, getNetworkSites, isConnectedSecondaryNetworkSite).
The selectors will be used as a replacement of current site-list methods.

**To test:**
npm run test-client client/state/selectors/test

### CODE REVIEW

**Comment by jorgefilipecosta** [comment]
Rebased

**Comment by gwwar** [comment]
@tyxla did we already have some selectors for these?

File: client/state/selectors/index.js [hunk 1]
```diff
@@ -162,14 +163,14 @@ export isFollowing from './is-following';
...
 export isFetchingPublicizeShareActionsPublished from './is-fetching-publicize-share-actions-published';
-export isSchedulingPublicizeShareAction from './is-scheduling-publicize-share-action';
 export isJetpackModuleActive from './is-jetpack-module-active';
 export isJetpackModuleUnavailableInDevelopmentMode from './is-jetpack-module-unavailable-in-development-mode';
 export isJetpackSettingsSaveFailure from './is-jetpack-settings-save-failure';
 export isJetpackSiteConnected from './is-jetpack-site-connected';
 export isJetpackSiteInDevelopmentMode from './is-jetpack-site-in-development-mode';
 export isJetpackSiteInStagingMode from './is-jetpack-site-in-staging-mode';
 export isJetpackUserMaster from './is-jetpack-user-master';
+export isMainSiteToSecondarySiteConnection from './is-main-site-to-secondary-site-connection';
```
Comment (tyxla) [inline client/state/selectors/index.js]: I really think we should find a better name for that one. Maybe @gwwar might have good ideas.

File: client/state/selectors/index.js [hunk 2]
```diff
@@ -134,6 +133,7 @@ export hasBrokenSiteUserConnection from './has-broken-site-user-connection';
 export hasInitializedSites from './has-initialized-sites';
 export hasSiteComments from './has-site-comments';
 export hasUnsavedUserSettings from './has-unsaved-user-settings';
+export hasUserAskedADirectlyQuestion from './has-user-asked-a-directly-question';
```
Comment (tyxla) [inline client/state/selectors/index.js]: I love that you fixed the order of selectors, but could we have done that in a separate PR?

File: client/state/selectors/get-network-sites.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ */
+import createSelector from 'lib/create-selector';
+import { isMainSiteToSecondarySiteConnection } from 'state/selectors';
+import { getSite } from 'state/sites/selectors';
+
+/**
+ * Returns the secondary sites of a given site
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {Number}  siteId      The ID of the site we're retrieving the secondary sites
```
Comment (tyxla) [inline client/state/selectors/get-network-sites.js]: Could make this one more clear? Maybe something like:

```
The ID of the main site for which we're retrieving the secondary sites
```

File: client/state/selectors/is-connected-secondary-network-site.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ */
+import createSelector from 'lib/create-selector';
+import { isMainSiteToSecondarySiteConnection } from 'state/selectors';
+
+/**
+ * Returns true if site with id equal to siteId is a connected secondary network site and false otherwise
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {Number}  siteId      The ID of the site we're querying
+ * @return {?Boolean}            Whether site with id equal to siteId is a connected secondary network site
```
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: `some` will always return a Boolean value, so let's remove the `null` notation from the JSdoc.

File: client/state/selectors/is-main-site-to-secondary-site-connection.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Internal dependencies
+ */
+import { getSiteOption, isJetpackSiteMainNetworkSite, isJetpackSiteSecondaryNetworkSite } from 'state/sites/selectors';
+import { withoutHttp } from 'lib/url';
+
+/**
+ * Returns true if site with id equal to mainSiteId is the main site of site with id equal to secondarySiteId and false otherwise.
```
Comment (tyxla) [inline client/state/selectors/is-main-site-to-secondary-site-connection.js]: Looks a little verbose, could we try to simplify? If we consider the argument names are clear enough to signify we're talking about site IDs, we could go with something like:

```
Returns true if site mainSiteId is the main site of site secondarySiteId and false otherwise.
```

File: client/state/selectors/is-main-site-to-secondary-site-connection.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * Returns null if with the information available in state the relationship is unknown.
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {Number}  mainSiteId      The ID of the main site
```
Comment (tyxla) [inline client/state/selectors/is-main-site-to-secondary-site-connection.js]: Can we align the JSdoc sections vertically?

File: client/state/selectors/get-network-sites.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * @return {?Array}            Array of secondary sites
```
Comment (tyxla) [inline client/state/selectors/get-network-sites.js]: The result array will never be `null`, so let's remove the `null` notation from the JSdoc.

File: client/state/selectors/is-main-site-to-secondary-site-connection.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * @param  {Number}  secondarySiteId      The ID of the main site
+ * @return {?Boolean}            Whether site with id equal to mainSiteId is main site of site with id equal to secondarySiteId
+ */
+export default ( state, mainSiteId, secondarySiteId ) => {
```
Comment (tyxla) [inline client/state/selectors/is-main-site-to-secondary-site-connection.js]: I like it that this selector will return `null` if we don't have any of those sites in the state. 

Would that make sense for the rest of the selectors that you're introducing here? What I mean specifically is:

* for `getNetworkSites`, return `null` if the site with the passed `id` is not found in the state.
* for `isConnectedSecondaryNetworkSite`, return `null` if any of the sites is not in the state.

This could be helpful when building UIs, so we know whether to show a placeholder or not.

**Comment by tyxla** [comment]
> @tyxla did we already have some selectors for these?

Nope, not for these.

**Comment by jorgefilipecosta** [comment]
Hi @tyxla, All the feedback was applied. Given that now the selectors are able to return null, a considerable amount of logic had to be added.

The most complex one is isConnectedSecondaryNetworkSite and the logic is the following:
If the site passed is not found return null.
If the site passed is not a secondary site return false.
If it is true that some site is the main site of the site passed return true.
If it is not possible to determine if some site is the main site of the site passed return null.
If none of the sites passed are the main site of the site passed return false;

Feel free to have a new look, and thank you a lot for the great feedback and suggestions.

**Comment by jorgefilipecosta** [comment]
Regarding the name of selector isMainSiteToSecondarySiteConnection a shorter alternative would be isMainSiteOf, let me know if you prefer this version, if yes I will make the change.

**Comment by tyxla** [comment]
`isMainSiteOf` looks interesting, clear but still short enough, I kinda like it and would prefer it to the longer one - @gwwar what do you think?

**Comment by gwwar** [comment]
Nice I like it. Let's go with that.

**Comment by jorgefilipecosta** [comment]
Thank you @gwwar, @tyxla the selector was renamed :)

File: client/state/selectors/get-network-sites.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * Returns null if main site is not found
+ *
+ * @param  {Object}  state     Global state tree
+ * @param  {Number}  siteId    The ID of the main site for which we're retrieving the secondary sites
+ * @return {?Array}            Array of secondary sites
+ */
+export default createSelector(
+	( state, siteId ) => {
+		const isMainNetworkSite = isJetpackSiteMainNetworkSite( state, siteId );
+		if ( isMainNetworkSite == null ) {
```
Comment (tyxla) [inline client/state/selectors/get-network-sites.js]: It's always recommended to use the strict comparison operator `===` instead of the equality one `==`.

File: client/state/selectors/get-network-sites.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+			return null;
+		}
+		if ( ! isMainNetworkSite ) {
```
Comment (tyxla) [inline client/state/selectors/get-network-sites.js]: I think this case is confusing. I'd probably expect it to return `null` if the site is not the main network site. Because this one can easily be confused with the case when there is a main network site without any connected secondary sites.

File: client/state/selectors/is-connected-secondary-network-site.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * Returns null if with the information available it's not possible to determine if site is a connected secondary site or not
+ *
+ * @param  {Object}    state     Global state tree
+ * @param  {Number}    siteId    The ID of the site we're querying
+ * @return {?Boolean}            Whether site with id equal to siteId is a connected secondary network site
+ */
+export default createSelector(
+	( state, siteId ) => {
+		const isSecondaryNetworkSite = isJetpackSiteSecondaryNetworkSite( state, siteId );
+		if ( isSecondaryNetworkSite == null ) {
```
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: It's always recommended to use the strict comparison operator `===` instead of the equality one `==`.

File: client/state/selectors/is-connected-secondary-network-site.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import createSelector from 'lib/create-selector';
+import { isMainSiteOf } from 'state/selectors';
+import { isJetpackSiteSecondaryNetworkSite } from 'state/sites/selectors';
+
+/**
```
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: I'm not sure I understand the use case this selector would be needed in - can you please explain why we need this selector and we can't just use just `isJetpackSiteSecondaryNetworkSite`?

Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla, the use case of this selector is when displaying website where a plugin is installed. If the website is connected to a main site you don't show it in the main list. If it is not connected and has the plugin installed it is shown. You can check the PR that makes use of this selector here: https://github.com/Automattic/wp-calypso/pull/15742/files.

**Comment by jorgefilipecosta** [comment]
Pull request rebased and feedback applied.

File: client/state/selectors/is-connected-secondary-network-site.js (hunk 12 above)
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: Gotcha, thanks for clarifying. Indeed there is quite a difference between a connected secondary site, and a connected secondary site where the main network site is also connected.

So, I have a suggestion: let's make sure to describe what's the purpose of this selector. It can be confusing to understand the difference between it and `isJetpackSiteSecondaryNetworkSite` at first glance without looking at the code.

Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla , you are right, with the comments it was  not possible to understand the difference of the selectors. So I added a clarification in the comments to make things easier. Thank you for the suggestion.

File: client/state/selectors/is-connected-secondary-network-site.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+		if ( isSecondaryNetworkSite === null ) {
+			return null;
+		}
+		if ( ! isSecondaryNetworkSite ) {
+			return false;
+		}
+		const siteIds = Object.keys( get( state, 'sites.items', {} ) );
+		return siteIds.reduce( ( previousResult, mainSiteId ) => {
+			if ( previousResult ) {
+				return previousResult;
```
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: This logic indicates we're probably not supposed to be using `reduce` here. How about using lodash's `some` instead of the `reduce`?

File: client/state/selectors/test/get-network-sites.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+							main_network_site: 'https://example.wordpress.com'
+						}
+					},
+				}
+			}
+		};
+		expect( getNetworkSites( state, 2 ) ).to.be.null;
+	} );
+
+	it( 'should return  null if site is not a main site', () => {
```
Comment (tyxla) [inline client/state/selectors/test/get-network-sites.js]: We have a double space here.

File: client/state/selectors/is-connected-secondary-network-site.js (hunk 13 above)
Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla, In this case to allow the selector to return null, we don’t want just to return true or false, as lodash some does. It would be possible to use lodash some 2 times, doing 2 iterations one that checks for true and other that checks for null  And if non is true we return false. Using reduce as it is right now we just need one iteration that’s why I used it. The version with lodash some may be easier to understand and If you prefer I can switch.

File: client/state/selectors/test/get-network-sites.js (hunk 14 above)
Comment (jorgefilipecosta) [inline client/state/selectors/test/get-network-sites.js]: Nice catch 👍 Code updated.

File: client/state/selectors/is-connected-secondary-network-site.js (hunk 13 above)
Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: I don't quite follow your 2 iterations explanation. What I mean is, replace:

```
return siteIds.reduce( ( previousResult, mainSiteId ) => {
	if ( previousResult ) {
		return previousResult;
	}
	const newResult = isMainSiteOf( state, mainSiteId, siteId );
	if ( newResult || newResult === null ) {
		return newResult;
	}
	return previousResult;
}, false );
```

with the much simpler one that uses `lodash` some:

```
return some( siteIds, ( mainSiteId ) => {
	const isMainSite = isMainSiteOf( state, mainSiteId, siteId );
	return isMainSite || isMainSite === null;
} );
```

Does that make sense?

Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla, The problem with this approach is that if some isMainSiteOf returns null and all the others false or null we will return true when we should be returning null.

Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: Which makes me think: why would we want to consider the `null` a truthful value? If we don't it would be even simpler:

```
some( siteIds, ( mainSiteId ) => isMainSiteOf( state, mainSiteId, siteId ) )
```

Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: In other words, if you want this selector to solve a different type of problem, it's worth adding a test for that problem. Current tests would pass with what I'm suggesting.

Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla, you are right one of my tests was not properly testing the case as it should and that was improved 👍 
The equivalent version with lodash some to my reduce code would be:
```
		if ( some( siteIds, ( mainSiteId ) => isMainSiteOf( state, mainSiteId, siteId ) ) ) {
			return true;
		}
		if ( some( siteIds, ( mainSiteId ) => isMainSiteOf( state, mainSiteId, siteId ) === null ) ) {
			return null;
		}
		return false;
```
The reason why this selector returns null its because its a common practice that selectors return null if its not possible to determine the output. In this case, it's not possible to determine if non of the isMainSiteOf returned true and at least one returned null. But until now there is no practical use of null in this selector and if you prefer I can make it return just true or false.

Comment (tyxla) [inline client/state/selectors/is-connected-secondary-network-site.js]: Thanks for clarifying @jorgefilipecosta.

To be honest, I think we can go without a `null` on this one - I can't think of a situation where a `null` here could be helpful. Let's go with that, and if we need the "not possible to determine" case, we can always iterate, right 😉

Comment (jorgefilipecosta) [inline client/state/selectors/is-connected-secondary-network-site.js]: Hi @tyxla, this selector now returns just true or false, I also think in this case we are totally fine with this approach.

**Comment by tyxla** [comment]
We're good to go here, great work @jorgefilipecosta! Did you want to do anything else here before we 🚢 ?

**Comment by jorgefilipecosta** [comment]
Thank you @tyxla, I had a new look and it seems ok, its a go for me :)
